improve chordMetadata thread safety

rhaas80 opened the pull request.

rhaas80 committed:
Buffer: use deepCopy to copy metadata

rhaas80 committed:
metadataObject: remove dummy function and make abstract base class

rhaas80 committed:
BasebandMetadata: add from / to json and deepCopy functions

rhaas80 committed:
HFBMetadata: add deepCopy function

rhaas80 committed:
BeamMetadata: add deepCopy function

rhaas80 committed:
N2Metadata: add to / from json functions

rhaas80 committed:
chordMetadata: add to / from json and deepCopy functions

rhaas80 committed:
CHORDTelescope: add EOP from / to json functions

rhaas80 committed:
chordMetadata: try and make deepCopy thread safe

rhaas80 committed:
metadataObject: make deepCopy take a const source

rhaas80 committed:
BasebandMetadat: adapt to deepCopy taking a const source

rhaas80 committed:
BeamMetadata: adapt to deepCopy taking a const source

rhaas80 committed:
visBuffer: adapt to deepCopy taking a const source

rhaas80 committed:
HFBMetadata: adapt to deepCopy taking a const source

rhaas80 committed:
chordMetadata: adapt to deepCopy taking a const source

rhaas80 committed:
N2Metadata: adapt to deepCopy taking a const source

rhaas80 committed:
NDArrayRingBuffer: use deepCopy

rhaas80 committed:
cudaCopyFromRingbuffer: use deepCopy

rhaas80 committed:
cudaCopyToRingbuffer: use deepCopy

rhaas80 committed:
NDArrayBuffer: use deepCopy

rhaas80 committed:
cudaFRBBeamReformer: use deepCopy

geoffryan commented on lib/metadata/BasebandMetadata.cpp:
The `from_json` should use the checked access `.at()` instead of unchecked `[]`.

geoffryan commented on lib/metadata/BeamMetadata.cpp:
As above, the `from_json` should use the checked access `.at()` instead of unchecked `[]`.

geoffryan commented on lib/metadata/HFBMetadata.cpp:
As above, the `from_json` should use the checked access `.at()` instead of unchecked `[]`.

geoffryan commented on lib/metadata/N2Metadata.cpp:
As above, the `from_json` should use the checked access `.at()` instead of unchecked `[]`.

geoffryan commented on lib/metadata/chordMetadata.hpp:
Does this need a `void to_json(nlohmann::json& j, const chordMetadata& m)` and `void from_json(const nlohmann::json& j, BasebandMetadata& m)` ?

Should the `to_json` include (perhaps temporarily) the non-json'd members?

rhaas80 replied:
it should, it must.

geoffryan commented on lib/utils/CHORDTelescope.cpp:
As above, the from_json should use the checked access .at() instead of unchecked [].

geoffryan reviewed (changes requested):
Looks good! Can't spot any problems with `deepCopy`.  A few comments on the `json`:

1. I think the `from_json` methods should use checked accesses `.at()`, unless there's a pressing reason not to.
2. The `to_json` (I believe) always constructs a fresh `json` (with no existing members), so the checked `emplace()` can be used instead of `[]`.
3. `chordMetadata` is missing to/from json functions.
4. While existing (and future) CI tests will cover the `deepCopy` functionality, I don't think the to/from json is covered by any testing.  I think the right place for this would be in `tests/boost`.  Could you add some quick tests there that ensure the metadata  types can survive the round trip through json?

rhaas80 committed:
BasebandMetadata: check access when (de)serializing with json

rhaas80 committed:
N2Metadata: check access when (de)serializing with json

rhaas80 committed:
EOP: check access when (de)serializing with json

rhaas80 committed:
BeamMetadata: check access when (de)serializing with json

rhaas80 committed:
HFBMetadata: check access when (de)serializing with json

rhaas80 committed:
chordMetadata: add to/from json functions

rhaas80 committed:
DataType: add from/to json functions

rhaas80 committed:
hdf5FileWrite: use const ref for input shared_ptr<const>

rhaas80 committed:
cudaCopyNToRingbuffer: chordMetadata->deepCopy

rhaas80 committed:
chordMetadata: add operator== for unit tests

rhaas80 committed:
boost-tests: update dependencies for json and Telescope in Metadata

rhaas80 committed:
jsonMetadata: add to_json for beamCoord

rhaas80 committed:
chordMetadata: add set_beam_coord member function

rhaas80 committed:
chordMetadata: (de)serialize all members checked by == to bytes

rhaas80 committed:
jsonMetadata: make beamCoord a plain-old-C struct

so that we can serialize and initialize it as bytes

rhaas80 committed:
test_metadataObject: add unit test for chordMetadata (de)serialization

rhaas80 commented:
> 1. I think the `from_json` methods should use checked accesses `.at()`, unless there's a pressing reason not to.

No reason why I should not use `.at()`. Fixed.

> 2. The `to_json` (I believe) always constructs a fresh `json` (with no existing members), so the checked `emplace()` can be used instead of `[]`.

Sounds good. Fixed.

> 3. `chordMetadata` is missing to/from json functions.

hmm, let's see. Oh right, it has the `to_json` member function (must have, it's pure-virtual in `metadataObject`) but no free-standing to/from json code. That will actually be some minutes of work since I also need to serialize / de-serialize that leftover structure members (eg the array description)....

> 4. While existing (and future) CI tests will cover the `deepCopy` functionality, I don't think the to/from json is covered by any testing.  I think the right place for this would be in `tests/boost`.  Could you add some quick tests there that ensure the metadata  types can survive the round trip through json?

Sure. Tests provided in commit 31176d7d1

That triggered a bunch of other changes since eg the to / from bytes serializations were also not quite complete yet. The unit test above checks both serialization to / from json and to / from bytes.

rhaas80 committed:
chordMetadata: pacify gcc stringop warning

rhaas80 committed:
chordMetadata: (de)serialize only json elements that exist

geoffryan reviewed (approved):
Fantastic!  LGTM

rhaas80 commented:
Very good. I'll merge just after #1327 (and once I've removed access to `name`).

rhaas80 commented on lib/metadata/chordMetadata.hpp:
```diff
@@ -46,6 +46,7 @@ const int CHORD_META_MAX_VIS_SAMPLES = 64;
 class chordMetadata : public metadataObject {
 public:
     chordMetadata();
+    bool operator==(const chordMetadata& other);
```
actually should this be `operator==() const`?

rhaas80 committed:
gpuSimulateN2kCorr: use chordMetadata::deepCopy

rhaas80 committed:
gpuSimulateN2kPL1bitCorr: use chordMetadata::deepCopy

rhaas80 committed:
gpuSimulateN2kPLExpand: use chordMetadata::deepCopy
